### PR TITLE
docs: update and expand 'Configuration files' section

### PR DESCRIPTION
**Pull request by elmjag**
Update and exapand documentation of YAML configuration files.

- remove outdated information
- update names of changed identifiers
- add more details and some examples

### CODE REVIEW

**Comment by fabcor-maxiv** [comment]
Preview here: https://mxcubecore--993.org.readthedocs.build/en/993/dev/configuration_files.html

File: docs/source/dev/configuration_files.md [hunk 1]
```diff
@@ -37,41 +37,84 @@ Each YAML-configured object has a `name` attribute,
 which is equal to the role that identifies the object within the containing object
 (the name of the Beamline object is `beamline`).
 
-YAML-configured objects must be subclasses of the `BaseHardwareObjects.ConfiguredObject` class.
-HardwareObjects proper (which excludes e.g. `Beamline` and procedures)
-are subclasses of `BaseHardwareObjects.HadwareObjectYaml`.
-The most complete example is the Beamline object and the comments in `beamline_config.yml`
-are the best guide to the syntax of YAML configuration files.
+The most complete example of an YAML-configured object is the Beamline object.
```
Comment (fabcor-maxiv) [inline docs/source/dev/configuration_files.md]:
```suggestion
The most complete example of a YAML-configured object is the `Beamline` object.
```

It is also possible to link directly to the API documentation for the `Beamline` class, but I never recall the correct notation, it needs to be looked up in [MyST parser documentation](https://myst-parser.readthedocs.io/) and tested.

File: docs/source/dev/configuration_files.md [hunk 2]
```diff
@@ -37,41 +37,84 @@ Each YAML-configured object has a `name` attribute,
...
 Note that classes are loaded and initialised in the order given by this list,
 so that there is a reproducible loading order.
 Contained objects can be defined as procedures, so that they are added to the list of procedures.
 Each YAML-configured class has an `_init()` method that is executed immediately after the object is created,
 and an `init()` function that is executed after configured parameters and contained objects have been loaded.
 
-### Accessing configuration data
+Below is an example YAML configuration file:
+
+```
```
Comment (fabcor-maxiv) [inline docs/source/dev/configuration_files.md]:
```suggestion
```yaml
```

File: docs/source/dev/configuration_files.md [hunk 3]
```diff
@@ -37,41 +37,84 @@ Each YAML-configured object has a `name` attribute,
...
+class: ISPyBClientMockup.ISPyBClientMockup
+configuration:
+  base_result_url: "https://your.limsresults.org"
```
Comment (fabcor-maxiv) [inline docs/source/dev/configuration_files.md]:
```suggestion
  base_result_url: https://your.limsresults.org
```

I think quotes are not necessary here, and yamllint would complain.

File: docs/source/dev/configuration_files.md [hunk 4]
```diff
@@ -37,41 +37,84 @@ Each YAML-configured object has a `name` attribute,
...
+  login_type: proposal
+objects:
+  lims_rest: lims_rest.yml
+  session: session.yml
```
Comment (fabcor-maxiv) [inline docs/source/dev/configuration_files.md]:
```suggestion
  session: session.yaml
```

File: docs/source/dev/configuration_files.md [hunk 5]
```diff
@@ -37,41 +37,84 @@ Each YAML-configured object has a `name` attribute,
...
+```
+
+This file specifies a hardware object, which is an instance of the `ISPyBClientMockup` class.
+That object will have two configuration properties `base_result_url` and `login_type`.
+Two child objects with roles `lims_rest` and `session` will be loaded from the specified configuration files.
+
+### Accessing configuration properties
+
+The contents of the `configuration` section will be available as an `config` attribute of the hardware object.
```
Comment (fabcor-maxiv) [inline docs/source/dev/configuration_files.md]:
```suggestion
The contents of the `configuration` section will be available as a `config` attribute of the hardware object.
```

File: docs/source/dev/configuration_files.md [hunk 6]
```diff
@@ -37,41 +37,84 @@ Each YAML-configured object has a `name` attribute,
...
+It is also possible to access the configuration with `get_property()` and `get_properties()` methods.
```
Comment (fabcor-maxiv) [inline docs/source/dev/configuration_files.md]:
```suggestion
It is also possible to access the configuration with the `get_property()` and `get_properties()` methods.
```

File: docs/source/dev/configuration_files.md [hunk 7]
```diff
@@ -37,41 +37,84 @@ Each YAML-configured object has a `name` attribute,
...
+Below is an example of how  configuration can be read in `init()` method of a hardware object.
```
Comment (fabcor-maxiv) [inline docs/source/dev/configuration_files.md]:
```suggestion
Below is an example of how configuration can be read in the `init()` method of a hardware object.
```

File: docs/source/dev/configuration_files.md [hunk 8]
```diff
@@ -37,41 +37,84 @@ Each YAML-configured object has a `name` attribute,
...
+
+```
```
Comment (fabcor-maxiv) [inline docs/source/dev/configuration_files.md]:
```suggestion
```python
```

File: docs/source/dev/configuration_files.md [hunk 9]
```diff
@@ -37,41 +37,84 @@ Each YAML-configured object has a `name` attribute,
...
+The values of an object's configuration properties are considered implementation details of that hardware object.
+If access is required by outside code, then it should be provided by the object's client API.
+
+### Accessing child objects
+
+The `objects` dictionary in the YAML configuration file specifies the child objects.
+These child objects can be accessed via the parent object's `role_name` attribute.
+For example, if a hardware object is configured with this configuration file:
+
+```
```
Comment (fabcor-maxiv) [inline docs/source/dev/configuration_files.md]:
```suggestion
```yaml
```

File: docs/source/dev/configuration_files.md [hunk 10]
```diff
@@ -37,41 +37,84 @@ Each YAML-configured object has a `name` attribute,
...
+class: Foo.Foo
+objects:
+  bar: gazonk.yml
```
Comment (fabcor-maxiv) [inline docs/source/dev/configuration_files.md]:
```suggestion
  bar: gazonk.yaml
```

File: docs/source/dev/configuration_files.md [hunk 11]
```diff
@@ -37,41 +37,84 @@ Each YAML-configured object has a `name` attribute,
...
+```
+
+An instance of the `Foo` class will be created.
+This instance will have a child object with `bar` role, with configuration from `gazonk.yml` file.
```
Comment (fabcor-maxiv) [inline docs/source/dev/configuration_files.md]:
```suggestion
This instance will have a child object with `bar` role, with configuration from `gazonk.yaml` file.
```

**Review by fabcor-maxiv — COMMENTED** [review]
Good stuff

File: docs/source/dev/configuration_files.md (hunk 2 above)
Comment (elmjag) [reply docs/source/dev/configuration_files.md]: Hah, cool!

File: docs/source/dev/configuration_files.md (hunk 8 above)
Comment (elmjag) [reply docs/source/dev/configuration_files.md]: Currently, this does not do anything. But it does not hurt to specify that it's python code :P

Comment (fabcor-maxiv) [reply docs/source/dev/configuration_files.md]: Sphinx's HTML renderer and/or the MyST parser is/are smart enough. But if we swap to some alternative components of the toolchain (say to render in Latex, or PDF for example), then maybe the syntax is not automatically detected.

**Review by fabcor-maxiv — APPROVED** [review]

**Review by rhfogh — APPROVED** [review]
Looks good - and really needed updating.

**Pull request merged**